libs/extensions/src/NanoEvent.cpp

```diff
@@ -31,4 +52,174 @@ shared_ptr<PhysicsObjects> NanoEvent::GetAllMuons(float matchingDeltaR)
   }
 
   return allMuons;
-}
\ No newline at end of file
+}
+
+shared_ptr<PhysicsObjects> NanoEvent::GetSegmentMatchedMuons(int minMatches)
+{
+  auto looseMuons = GetCollection("LooseMuons");
+  auto looseDsaMuons = GetCollection("LooseDSAMuons");
+
+  bool useMatchRatio = false;
+  if(minMatches == 0) useMatchRatio=true;
+
+  auto allMuons = make_shared<PhysicsObjects>();
+  for(auto muon : *looseMuons){
+    allMuons->push_back(muon);
+  }
+  for(auto dsaMuon : *looseDsaMuons){
+
+    if(useMatchRatio) {
+      float nHits = float(dsaMuon->Get("trkNumDTHits")) + float(dsaMuon->Get("trkNumCSCHits"));
+      minMatches = static_cast<int>(nHits * 2/3);
+    }
+    
+    bool matchFound = false;
+    if(float(dsaMuon->Get("muonMatch1")) >= minMatches) {
+      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch1idx")),true) matchFound = true;
+    }
+    if(!matchFound && float(dsaMuon->Get("muonMatch2")) >= minMatches) {
+      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch2idx")),true) matchFound = true;
+    }
+    if(!matchFound && float(dsaMuon->Get("muonMatch3")) >= minMatches) {
+      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch3idx")),true) matchFound = true;
+    }
+    if(!matchFound && float(dsaMuon->Get("muonMatch4")) >= minMatches) {
+      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch4idx")),true) matchFound = true;
+    }
+    if(!matchFound && float(dsaMuon->Get("muonMatch5")) >= minMatches) {
+      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch5idx")),true) matchFound = true;
+    }
+    if(matchFound == false) allMuons->push_back(dsaMuon);
+
+  }
+
+  return allMuons;
+}
+
+bool NanoEvent::IndexExist(shared_ptr<PhysicsObjects> objectCollection, float index, bool isDSAMuon) {
+  for(auto object : *objectCollection) {
+    if(float(object->Get("idx")) == index) {
+      if(isDSAMuon && asNanoMuon(object)->isDSAMuon()) {
+        return true;
+      }
+      if(!isDSAMuon && !asNanoMuon(object)->isDSAMuon()) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+float NanoEvent::DeltaR(float eta1, float phi1, float eta2, float phi2) {
+  float dEta = eta1 - eta2;
+  float dPhi = TVector2::Phi_mpi_pi(phi1-phi2);
+  return TMath::Sqrt(dEta*dEta + dPhi*dPhi);
+}
+
+shared_ptr<PhysicsObjects> NanoEvent::GetAllMuonVertexCollections() {
```

Comment:
I'm wondering if this is the best name, since it returns 1 collection. Maybe it should be GetAllMuonVerticesCollection()?